File: state.py
"""Explicit conversational facts. Copy on update so failed calls don't commit state."""
import re


def create_state():
    return {"pendingQuestion": None, "daysSincePurchase": None, "isPremiumMember": None, "orderNumber": None}


def apply_reply(previous, question):
    state = dict(previous)
    reply = question.strip()
    if state["pendingQuestion"] == "order_number" and re.fullmatch(r"[0-9]+", reply):
        state["orderNumber"] = reply
        state["pendingQuestion"] = None
        return state
    days = re.findall(r"\b([0-9]+) days?\b", reply, re.IGNORECASE)
    if days:
        state["daysSincePurchase"] = int(days[-1])
    elif state["pendingQuestion"] == "purchase_age" and re.fullmatch(r"[0-9]+", reply):
        state["daysSincePurchase"] = int(reply)
    membership = re.findall(r"\b(not a premium|a premium|standard) member\b", reply.lower())
    if membership:
        state["isPremiumMember"] = membership[-1] == "a premium"
    elif state["pendingQuestion"] == "membership":
        if re.fullmatch(r"yes[.!]?", reply, re.IGNORECASE):
            state["isPremiumMember"] = True
        if re.fullmatch(r"no[.!]?", reply, re.IGNORECASE):
            state["isPremiumMember"] = False
    return state


def record_next_question(state, decision):
    return {**state, "pendingQuestion": {"ask_days": "purchase_age", "ask_membership": "membership"}.get(decision)}

File: typescript/state.ts
import type { ConversationState, Decision } from "./types.ts";

export function createState(): ConversationState {
  return { pendingQuestion: null, daysSincePurchase: null, isPremiumMember: null, orderNumber: null };
}

// Return a copy: a failed API call must not commit half a conversation turn.
export function applyReply(previous: ConversationState, question: string): ConversationState {
  const state = { ...previous };
  const reply = question.trim();
  if (state.pendingQuestion === "order_number" && /^\d+$/.test(reply)) {
    state.orderNumber = reply;
    state.pendingQuestion = null;
    return state;
  }
  const days = [...reply.matchAll(/\b(\d+) days?\b/gi)].at(-1)?.[1];
  if (days !== undefined) state.daysSincePurchase = Number(days);
  else if (state.pendingQuestion === "purchase_age" && /^\d+$/.test(reply)) {
    state.daysSincePurchase = Number(reply);
  }
  const membership = [...reply.toLowerCase().matchAll(/\b(not a premium|a premium|standard) member\b/g)].at(-1)?.[1];
  if (membership !== undefined) state.isPremiumMember = membership === "a premium";
  else if (state.pendingQuestion === "membership") {
    if (/^yes[.!]?$/i.test(reply)) state.isPremiumMember = true;
    if (/^no[.!]?$/i.test(reply)) state.isPremiumMember = false;
  }
  return state;
}

export function recordNextQuestion(state: ConversationState, decision: Decision): ConversationState {
  return { ...state, pendingQuestion: decision === "ask_days" ? "purchase_age" : decision === "ask_membership" ? "membership" : null };
}
